### PR TITLE
add 4. selenium version to get compatibility with Selenium Grid 4.

### PR DESCRIPTION
**Pull request by alvarolaserna**
Right now, using selenium grid 4.14.1 Testray doesnt work, adding this upgrade will make it work.

### CODE REVIEW

**Comment by TDL-EdgarsEglitis** [comment]
I think this code can be simplified. It looks like `Selenium::WebDriver::Remote::Capabilities` is deprecated, so we could try something like this:
```ruby
  def build_chrome_driver(chrome_ops)
    chromeOptions = Selenium::WebDriver::Chrome::Options.new(
      options: chrome_ops,
    )
    if @url.nil?
      # local selenium instance
      driver = Selenium::WebDriver.for(
        :chrome, options: chromeOptions,
      )
    else
      # remote selenium grid
      log_debug("Selenium Server URL: #{@url}")
      driver = Selenium::WebDriver.for(
        :remote, url: @url, options: chromeOptions,
      )
    end
    return driver
  end
```

The equivalent code for other browsers should be similarly adjusted.

**Comment by alvarolaserna** [comment]
> I think this code can be simplified. It looks like `Selenium::WebDriver::Remote::Capabilities` is deprecated, so we could try something like this:
> 
> ```ruby
>   def build_chrome_driver(chrome_ops)
>     chromeOptions = Selenium::WebDriver::Chrome::Options.new(
>       options: chrome_ops,
>     )
>     if @url.nil?
>       # local selenium instance
>       driver = Selenium::WebDriver.for(
>         :chrome, options: chromeOptions,
>       )
>     else
>       # remote selenium grid
>       log_debug("Selenium Server URL: #{@url}")
>       driver = Selenium::WebDriver.for(
>         :remote, url: @url, options: chromeOptions,
>       )
>     end
>     return driver
>   end
> ```
> 
> The equivalent code for other browsers should be similarly adjusted.

I've modified it to cover that, the only one that Im not sure how to test is Safari, it doesnt seem to have any capabilities that I can try and work with... but for the rest of the browsers, everything goes as expected

**Comment by TDL-EdgarsEglitis** [comment]
Ok great - I can check it once you push the commit :)

**Comment by alvarolaserna** [comment]
> Ok great - I can check it once you push the commit :)

done

File: lib/core/device_drivers.rb [hunk 1]
```diff
@@ -179,22 +179,19 @@ def merge_chrome_ops(config_caps, case_caps)
 
   # build the Chrome driver, given a set of options
   def build_chrome_driver(chrome_ops)
+    # local selenium instance
+    chromeOptions = Selenium::WebDriver::Remote::Capabilities.chrome(
```
Comment (TDL-EdgarsEglitis) [inline lib/core/device_drivers.rb:183]: Why are these not `Selenium::WebDriver::Chrome::Options` like in my example? Like I said, `Selenium::WebDriver::Remote::Capabilities` is deprecated and should be removed.

**Review by TDL-EdgarsEglitis — CHANGES_REQUESTED** [review]
We should only use `Selenium::WebDriver::<browser>::Options.new`

File: lib/core/device_drivers.rb [hunk 2]
```diff
@@ -179,22 +179,19 @@ def merge_chrome_ops(config_caps, case_caps)
...
+      "goog:chromeOptions": chrome_ops,
+    )
     if @url.nil?
-      # local selenium instance
-      localChromeOptions = Selenium::WebDriver::Chrome::Options.new(
-        options: chrome_ops,
-      )
       driver = Selenium::WebDriver.for(
-        :chrome, options: localChromeOptions
+        :chrome, capabilities: chromeOptions
```
Comment (TDL-EdgarsEglitis) [inline lib/core/device_drivers.rb:188]: Not sure why this is `capabilities` instead of `options`

File: lib/core/device_drivers.rb [hunk 3]
```diff
@@ -282,19 +272,16 @@ def merge_edge_ops(config_caps, case_caps)
...
+        :edge, options: edgeOptions
       )
     else
       # remote selenium grid
       log_debug("Selenium Server URL: #{@url}")
-      remoteEdgeOptions = Selenium::WebDriver::Remote::Capabilities.edge(
-        "ms:edgeOptions" => edge_ops,
-      )
       driver = Selenium::WebDriver.for(
         :remote, url: @url, desired_capabilities: remoteEdgeOptions,
```
Comment (TDL-EdgarsEglitis) [inline lib/core/device_drivers.rb:286]: This needs to have `options: edgeOptions` instead of `desired_capabilities`

File: lib/core/device_drivers.rb [hunk 4]
```diff
@@ -257,21 +250,18 @@ def merge_ie_ops(config_caps, case_caps)
...
       )
     else
       # remote selenium grid
       log_debug("Selenium Server URL: #{@url}")
-      remoteIEOptions = Selenium::WebDriver::Remote::Capabilities.ie(
-        "se:ieOptions" => ie_ops,
-      )
       driver = Selenium::WebDriver.for(
-        :remote, url: @url, desired_capabilities: remoteIEOptions,
+        :remote, url: @url, desired_capabilities: iEOptions,
```
Comment (TDL-EdgarsEglitis) [inline lib/core/device_drivers.rb:264]: `options` instead of `desired_capabilities`

**Comment by TDL-EdgarsEglitis** [comment]
I just realised that the code can be simplified even further:
```ruby
  def build_chrome_driver(chrome_ops)
    chromeOptions = Selenium::WebDriver::Chrome::Options.new(
      options: chrome_ops,
    )
    if @url.nil?
      # local selenium instance
      Selenium::WebDriver.for(
        :chrome, options: chromeOptions,
      )
    else
      build_remote_driver(chromeOptions)
    end
  end
  
  # similar methods for other browser drivers
  
  def build_remote_driver(options)
    # remote selenium grid
    log_debug("Selenium Server URL: #{@url}")
    Selenium::WebDriver.for(
      :remote, url: @url, options: options,
    )
  end
  ```

File: lib/core/device_drivers.rb (hunk 1 above)
Comment (alvarolaserna) [reply lib/core/device_drivers.rb]: The way of adding prefs and args is a bit... well, not as simple, but I have changed it

File: lib/core/device_drivers.rb (hunk 2 above)
Comment (alvarolaserna) [reply lib/core/device_drivers.rb]: my mistake, fixed

File: lib/core/device_drivers.rb (hunk 1 above)
Comment (TDL-EdgarsEglitis) [reply lib/core/device_drivers.rb]: Checked the latest changes. If the `chrome_ops`/`firefox_ops` etc. are only expected to have the `args`/`prefs`/`profile` keys, it can be simplified with the splat operator:
```ruby
ops = Selenium::WebDriver::Chrome::Options.new(**chrome_opts)
 ```
 [See here for more details about it](https://ruby-doc.org/core-2.0.0/doc/syntax/calling_methods_rdoc.html#label-Hash+to+Keyword+Arguments+Conversion).

**Comment by alvarolaserna** [comment]
> ```ruby
>   def build_remote_driver(options)
>     # remote selenium grid
>     log_debug("Selenium Server URL: #{@url}")
>     Selenium::WebDriver.for(
>       :remote, url: @url, options: options,
>     )
>   end
> ```

Yes, this makes a lot of sense, thanks!

File: lib/core/device_drivers.rb (hunk 1 above)
Comment (alvarolaserna) [reply lib/core/device_drivers.rb]: Yeah, good point, Ill add that for all of them, it makes more sense

Comment (alvarolaserna) [reply lib/core/device_drivers.rb]: Just commited a change that solves this and adds the remote driver helper definition

File: examples/tests/cases/case_av_example.yaml [hunk 5]
```diff
@@ -48,7 +48,6 @@ TestBrowserSafariAV:
     - Role: desktopSafari
       Capabilities:
         safariOptions:
-          prefs:
```
Comment (TDL-EdgarsEglitis) [inline examples/tests/cases/case_av_example.yaml:51]: Was this change intentional? Does it not work if we use the splat operator when building the driver, like for other browsers?

Comment (alvarolaserna) [reply examples/tests/cases/case_av_example.yaml:51]: in the docs: https://www.rubydoc.info/gems/selenium-webdriver/Selenium/WebDriver/Safari/Options

there is not a constructor, so, I can pass it, but it will be useless... it might be from the parent Options, but I think right now is not implemented in any meaningful way.

Comment (TDL-EdgarsEglitis) [reply examples/tests/cases/case_av_example.yaml:51]: Thanks for the link. I agree that it seems there is no implementation for passing arguments, but I think it should still be fine to use the same splat for the Safari driver (unless it breaks something). All other browser constructors extend the parent anyway, so there shouldn't be much of a difference.
Do keep the TODO comment though :)

Comment (alvarolaserna) [reply examples/tests/cases/case_av_example.yaml:51]: Done, should be all good now

**Review by TDL-EdgarsEglitis — APPROVED** [review]
